### PR TITLE
add peek(io, UInt8) and peek(io, Char) (#16025)

### LINKED ISSUE #16025: modify Base.peek to return a Char and export it
The `peek` function is very useful and an efficient alternative to marking and resetting a stream when all you need to do is get a single character. The current `peek` returns a byte, however, which leads to errors when people assume that they can compare that byte to a character and check for a character (only works for UTF-8 streams and ASCII characters). This was a significant annoyance in implementing https://github.com/JuliaLang/julia/pull/16024. This is technically not breaking since `Base.peek` isn't exported. Most code that uses it will actually continue to work.

### PR DESCRIPTION
**Pull request by bicycle1885**
This pull request adds and exports `peek` from `Base`, and will close #16025. This works on `AbstractIOBuffer` and `IOStream`; more generic definition will be possible if the IO type supports `seek`.

### CODE REVIEW

File: test/file.jl [hunk 1]
```diff
@@ -1095,3 +1095,17 @@ function test_13559()
...
 end
 @unix_only test_13559()
+
+# test peek (#16025)
+let n = tempname()
+    w = open(n, "a")
+    io = open(n)
+    write(w, "A"); flush(w)
+    @test peek(io, UInt8) === 0x41
+    @test peek(io, Char) === 'A'
```
Comment (nalimilan) [inline test/file.jl]: Could you also test with a non-ASCII char?

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -5923,6 +5923,13 @@ Open a file and read its contents. `args` is passed to `read`: this is equivalen
 read(filename, args...)
 
 """
+    peek(stream::IO, T)
```
Comment (nalimilan) [inline base/docs/helpdb/Base.jl]: While you're at it, you could move the docstring inline.

**Comment by nalimilan** [comment]
Thanks, looks good! You should run `make julia-genstdlib` to update the docs. You should also add `peek` to `doc/stdlib/io-network.rst` (and maybe mention it in the manual too).

**Comment by nalimilan** [comment]
The tests failure is due to `priorityqueue.jl` having `using Base.Collections`, and `using Base.Collections` having a `peek` function. Not sure how to best handle that. Note that it also has an `eval` function; so that's not a new issue.

**Comment by laborg** [comment]
This has been implemented. See https://github.com/JuliaLang/julia/pull/28811

On Julia 1.7.1:
```
julia> using Test

julia> let n = tempname()
           w = open(n, "a")
           io = open(n)
           write(w, "A"); flush(w)
           @test peek(io, UInt8) === 0x41
           @test peek(io, Char) === 'A'
           read(io, UInt8)
           @test_throws EOFError peek(io, UInt8)
           @test_throws EOFError peek(io, Char)
           close(io); close(w)
           rm(n)
       end
```
succeeds.

**Pull request closed without merging**